onnxscript/autocast.py

```diff
@@ -1,30 +1,113 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
 from __future__ import annotations
 
 from typing import Any, Callable, Optional
 
 import numpy as np
+import onnx
+from onnx import helper, numpy_helper
 from onnx.defs import OpSchema
 
 from onnxscript import tensor, values
 
+# Conversions from python values to ONNX are used by both the script converter as well
+# as the eager-mode runtime and both need to be consistent. The script converter converts
+# python values into ONNX TensorProto, while the runtime converts python values into
+# ONNXScript runtime's value-representation (based on Tensor).
+
+
+# Utilities to convert a python value to TensorProto (for use by the script converter)
+
 
-def cast_scalar_to_tensor(x, dtype=None):
-    """Promotes scalar values (bool, int, float) into tensors of rank zero.
-    The optional argument dtype specifies the desired np.dtype of the tensor.
-    Other types of inputs are returned as is.
+def _py_type_to_onnx_type(pytype: type):
+    if pytype is bool:
+        return onnx.TensorProto.BOOL
+    if pytype is int:
+        return onnx.TensorProto.INT64
+    if pytype is float:
+        return onnx.TensorProto.FLOAT
+    if pytype is str:
+        return onnx.TensorProto.STRING
+    raise ValueError(f"Tensor element of type {pytype} not supported")
+
+
+def pyvalue_to_onnx_tensor(tensor_name: str, pyvalue):
+    if isinstance(pyvalue, np.ndarray):
+        return numpy_helper.from_array(pyvalue, tensor_name)
+    if isinstance(pyvalue, list):
```

Comment:
Check for `isinstance(pyvalue, Sequence)`?

[typing.Sequence](https://docs.python.org/3/library/typing.html#typing.Sequence)

Reply:
I think tuples are treated differently as attributes so

Reply:
May be ok to start by being more restrictive for now, IMO. It's a bit tricky overall since we want to handle things consistently in the static (converter) mode and the eager mode, and in the eager mode we (currently) lack some of the static information. Still, I think tuples should be doable, but perhaps we can do that in a separate PR.